src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSProxyContext.cs

```diff
@@ -0,0 +1,609 @@
+        private nint NextJSVHandle = -2;
+        private readonly List<nint> JSVHandleFreeList = new();
+
+#if !FEATURE_WASM_THREADS
+        private JSProxyContext()
+        {
+        }
+#else
+        public nint NativeTID;
+        public int ManagedTID;
+        public bool IsMainThread;
+        public JSSynchronizationContext SynchronizationContext;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public bool IsCurrentThread()
+        {
+            return ManagedTID == Thread.CurrentThread.ManagedThreadId;
+        }
+
+        [UnsafeAccessor(UnsafeAccessorKind.Field, Name = "thread_id")]
+        private static extern ref long GetThreadNativeThreadId(Thread @this);
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static IntPtr GetNativeThreadId()
+        {
+            return (int)GetThreadNativeThreadId(Thread.CurrentThread);
+        }
+
+        public JSProxyContext(bool isMainThread, JSSynchronizationContext synchronizationContext)
+        {
+            SynchronizationContext = synchronizationContext;
+            Interop.Runtime.InstallWebWorkerInterop();
+            NativeTID = GetNativeThreadId();
+            ManagedTID = Thread.CurrentThread.ManagedThreadId;
+            IsMainThread = isMainThread;
+        }
+#endif
+
+        #region Current operation context
+
+#if !FEATURE_WASM_THREADS
+        public static readonly JSProxyContext MainThreadContext = new();
+        public static JSProxyContext CurrentThreadContext => MainThreadContext;
+        public static JSProxyContext CurrentOperationContext => MainThreadContext;
+        public static JSProxyContext PushOperationWithCurrentThreadContext()
+        {
+            // in single threaded build we don't have to keep stack of operations and the context/thread is always the same
+            return MainThreadContext;
+        }
+#else
+
+        // Context of the main thread
+        private static JSProxyContext? _MainThreadContext;
+        public static JSProxyContext MainThreadContext
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get => _MainThreadContext!;
+            set => _MainThreadContext = value;
+        }
+
+        private sealed class PendingOperation
+        {
+            public JSProxyContext? CapturedContext;
+            public bool Called;
+            public bool Multiple;
+        }
+
+        [ThreadStatic]
+        private static List<PendingOperation>? _OperationStack;
+        private static List<PendingOperation> OperationStack
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get
+            {
+                return _OperationStack ??= new();
+            }
+        }
+
+        public static void PushOperationUnknownContext()
+        {
+            var stack = OperationStack;
+            // for SchedulePopOperation()
+            if (stack.Count > 0 && stack[stack.Count - 1].Called)
```

Comment:
Can we move this stack cleanup operation to a separate static method, since it's duplicated across multiple methods here?
Also, should it be a loop in case there are multiple scheduled pops? Or is that not possible

Reply:
Multiple should not be possible, but we could assert against it.